content/posts/2023-11-20-tools-for-learning-llvm-tablegen.md

```diff
@@ -0,0 +1,500 @@
+it at some point.
+
+Which might be a problem as TableGen only exists within LLVM. Unlike a language
+such as C++, TableGen does not have an array of resources.
+
+So, as well as joining a new project, you also now need to learn a new
+Domain Specific Language (DSL). You didn’t come to LLVM to learn a DSL, you
+probably came here to write a compiler.
+
+I can’t say when this problem might be solved, but the situation isn’t as
+bleak as it appears. There have been big improvements in TableGen tools
+recently, which means you can put more of your energy into the goals that
+brought you to LLVM in the first place.
+
+# A Brief Introduction to TableGen
+
+Imagine you wanted to represent the registers of an architecture. I’m going to
+use Arm's AArch64 in particular here.
+
+You could describe them in TableGen as:
+```bash
+$ cat register.td
+
+class Register<int _size, string _alias=""> {
+   int size = _size;
+   string alias = _alias;
+}
+
+// 64 bit general purpose registers are X<N>.
+def X0: Register<8> {}
+// Some have special alternate names.
+def X29: Register<8, "frame pointer"> {}
+// ...a lot more registers omitted.
+```
+
+By default, the TableGen compiler `llvm-tblgen` creates “records” - which are
+shown below.
+
+```bash
+$ ./bin/llvm-tblgen register.td
+
+------------- Classes -----------------
+class Register<int Register:_size = ?, string Register:_alias = ""> {
+ int size = Register:_size;
+ string alias = Register:_alias;
+}
+------------- Defs -----------------
+def X0 {        // Register
+ int size = 8;
+ string alias = "";
+}
+def X29 {       // Register
+ int size = 8;
+ string alias = "frame pointer";
+}
+```
+
+This is the intermediate representation (IR) of the TableGen compiler, similar
+to LLVM's LLVM IR.
+
+When using LLVM you would select a "target" which is the processor architecture
+you want to generate instructions for. TableGen's equivalent is a "backend".
+These backends do not generate instructions, but instead output a format for
+that backend's specific use case.
+
+For example, there is a backend that generates C++ code for 
+[searching](https://godbolt.org/z/5c696j1f9) data tables. Other examples are 
+C header files and [reStructuredText](https://docutils.sourceforge.io/rst.html)
+documentation.
+
+```
+                       TableGen source
+                               |
+ +--llvm-tblgen----------------|------------------------+
+ |                             v                        |
+ |              +----- Expanded records ----+           |
+ |              |                           |           |
+ |              v                           v           |
+ | +-------------------------+    +-------------------+ |
+ | | --gen-searchable-tables |    | Other backends... | |
+ | +-------------------------+    +-------------------+ |
+ |             |                            |           |
+ +-------------|----------------------------|-----------+
+               v                            v
```

Comment:
```suggestion
 |              |                           |           |
 +--------------|---------------------------|-----------+
                v                           v
```

Reply:
I really didn't want to point the misalignment of the input and output arrow, but it hurts my inner ASCII artist. :)

Reply:
Fixed, thanks!